### ISSUE TITLE
[feature] support normalization, scale and margin in MEVO

### ISSUE THREAD

**Issue by gongbudaizhe** [facebookresearch/fairscale#864]
## 🚀 Feature
currently, vanilla FC + softmax + nllloss is supported in MEVO. In face recognition, cosface loss is commonly used which additionally normalize both the input feature and fc weights and add scale and margin to the logits.

## Motivation
face recognition models are typically trained with millions of IDs. MEVO will have broader impact in face recognition field if CosFace is supported.

## Additional context
CosFace paper: https://arxiv.org/abs/1801.09414

**Comment by min-xu-ai** [comment]
wow, this is very interesting. I haven't seen this before. Thanks @gongbudaizhe. Do you have some example code of cosface loss that's easily understandable? If you have time, I can help you with that. Otherwise, I will try to find time with this in the future depending on my other workloads.

**Comment by gongbudaizhe** [comment]
Thanks for your interest. @min-xu-ai  Below is the sample code of cosface loss.

```python
x = backbone(img)  # output feature, (B, D), B batch size, D feature dim 
w = Tensor(N, D) # the weights of FC Layer, N num class, D feature dim.

# normalize feature and fc layer before multiplication
x = nn.functional.normalize(x, dim=1)
w = nn.functional.normalize(w, dim=1)
logits = torch.einsum('nc,kc->nk', [x, w])

# add margin and scale. They are hyperparameters.
margin, scale = 0.35, 64

# add margin
row_ind = torch.arange(x.shape[0], dtype=torch.long).to(x.device)
col_ind = im_label
logits[row_ind, col_ind] -= margin

# add scale
logits *= scale

loss = F.cross_entropy(logits, im_label)
```

**Comment by min-xu-ai** [comment]
Thanks @gongbudaizhe! Can you provide shape and dtype for x, w, im_label for big scale use cases for the code above?

**Comment by gongbudaizhe** [comment]
@min-xu-ai updated in my last comment : )

**Comment by min-xu-ai** [comment]
Is the 8-64 a reasonable size for B and 128-1024 a good number for D?

**Comment by gongbudaizhe** [comment]
> Is the 8-64 a reasonable size for B and 128-1024 a good number for D?

yes, these settings are good. Personally, I use B = 64, D = 256.

**Comment by min-xu-ai** [comment]
I have one more question @gongbudaizhe.

Is this needed in your research or you are doing a product with this? The reason I am asking is that this has impact on the use cases of face rec. Will this enable massive surveillance of people without their consent? We all need to be mindful about social impact of the technologies that we work on. Thank you in advance for the discussion.

**Comment by gongbudaizhe** [comment]
It is a good question. But I don't have a definitive answer.

Whether it is needed in my research or product doesn't really matter here. Once this feature is avaible, people will use it for various purposes.  Pens can kill, guns can protect.

CosFace loss is not only used in face recognition. It is also used in large scale image retrieval.

**Comment by min-xu-ai** [comment]
>  It is also used in large scale image retrieval.

Thank you. Can you provide some references for the above?

**Comment by gongbudaizhe** [comment]
CosFace used in landmark retrieval:
https://arxiv.org/pdf/1906.04087.pdf

ArcFace (a cross entropy variant which resembles to cosface), used in a more recent framework:
https://arxiv.org/pdf/2111.00775.pdf

**Comment by min-xu-ai** [comment]
HI @gongbudaizhe, sorry for the delay for getting to this. I have a PR that should add LMCL to MEVO. CosFace was a fun paper to read and I am really glad that it is highly cited and seems to be well-received.

I don't have any way to test it with model convergence. My testing included a unit test that matches MEVO's numerical output with the baseline implementation (thanks for providing it above!). I also verified that the memory usage of it is the same as non-LMCL case. I have NOT checked its runtime carefully but it seems similar to the non-LMCL MEVO.

Please let me know if you have further comments or suggestions. Thanks as always.
